File: 2024/Python/15.py
data = open("../inputs/15.txt").read().split("\n\n")

grid, moves = data

grid = [list(line) for line in grid.splitlines()]

H = len(grid)
W = len(grid[0])


def in_bounds(y, x):
    return 0 <= y < H and 0 <= x < W


moves = "".join(move.strip() for move in moves)

move_dirs = {"^": (-1, 0), ">": (0, 1), "<": (0, -1), "v": (1, 0)}

for y, line in enumerate(grid):
    for x, item in enumerate(line):
        if item == "@":
            start_pos = (y, x)


def move_robot(grid, pos, move):
    dy, dx = move_dirs[move]
    r_y, r_x = pos

    i = 0
    dir_items = []
    while True:
        dir_items.append(grid[r_y + i * dy][r_x + i * dx])
        if dir_items[-1] == "#" or grid[r_y + i * dy][r_x + i * dx] == ".":
            break
        i += 1

    if dir_items[-1] == "#":
        return grid, pos

    for j in range(len(dir_items)):
        grid[r_y + j * dy][r_x + j * dx] = dir_items[(j - 1) % len(dir_items)]

    return grid, (r_y + dy, r_x + dx)


pos = start_pos
for move in moves:
    grid, pos = move_robot(grid, pos, move)

print(
    sum(
        100 * y + x
        for y, line in enumerate(grid)
        for x, item in enumerate(line)
        if item == "O"
    )
)

wide_map = {"#": ["#", "#"], ".": [".", "."], "O": ["[", "]"], "@": ["@", "."]}

new_grid = []
for line in data[0].splitlines():
    ln = []
    for char in line:
        ln.extend(wide_map[char])
    new_grid.append(ln)

for y, line in enumerate(new_grid):
    for x, item in enumerate(line):
        if item == "@":
            start_pos = (y, x)

pos = start_pos

horiz = [(0, 1), (0, -1)]


def search(grid, pos, move):
    dy, dx = move_dirs[move]
    y, x = pos
    visited = set()

    def dfs(y, x):
        if not (0 <= y < len(grid) and 0 <= x < len(grid[0])):
            return True, set()
        if grid[y][x] == "#":
            return True, set()
        if grid[y][x] == ".":
            return False, {(y, x)}
        if (y, x) in visited:
            return False, set()

        visited.add((y, x))
        to_move = set()

        if (dy, dx) in horiz:
            stopped, positions = dfs(y + dy, x + dx)
            if stopped:
                return True, set()
            to_move.add((y, x))
            to_move.update(positions)
        else:
            if grid[y][x] == "[":
                stopped, right_pos = dfs(y, x + 1)
                if stopped:
                    return True, set()
                to_move.update(right_pos)
            elif grid[y][x] == "]":
                stopped, left_pos = dfs(y, x - 1)
                if stopped:
                    return True, set()
                to_move.update(left_pos)

            stopped, next_pos = dfs(y + dy, x)
            if stopped:
                return True, set()

            to_move.add((y, x))
            to_move.update(next_pos)

        return False, to_move

    next_y, next_x = y + dy, x + dx
    stopped, positions = dfs(next_y, next_x)
    if stopped:
        return False, set()
    return True, positions


def move_two(grid, pos, move):
    dy, dx = move_dirs[move]
    ry, rx = pos
    can_move, to_move = search(grid, pos, move)
    if not can_move:
        return grid, pos
    positions = list(to_move)
    if dx > 0:
        positions.sort(key=lambda p: -p[1])
    elif dx < 0:
        positions.sort(key=lambda p: p[1])
    elif dy > 0:
        positions.sort(key=lambda p: -p[0])
    else:
        positions.sort(key=lambda p: p[0])
    for y, x in positions:
        if (y - dy, x - dx) in to_move:
            grid[y][x] = grid[y - dy][x - dx]
        else:
            grid[y][x] = "."

    grid[ry][rx] = "."
    grid[ry + dy][rx + dx] = "@"

    return grid, (ry + dy, rx + dx)


for move in moves:
    new_grid, pos = move_two(new_grid, pos, move)

print(
    sum(
        100 * y + x
        for y, line in enumerate(new_grid)
        for x, item in enumerate(line)
        if item == "["
    )
)

File: 2024/Deno/15.ts
const data = (await Deno.readTextFile("../inputs/15.txt")).split("\n\n")

const [grid, moves] = data

type Grid = string[][]
type Position = [number, number]
type Direction = [-1, 0] | [1, 0] | [0, -1] | [0, 1]

const gridArr: Grid = grid.split("\n").map(line => [...line])

const H = gridArr.length
const W = gridArr[0].length

const inBounds = (y: number, x: number): boolean => 
  0 <= y && y < H && 0 <= x && x < W

const movesStr = moves.replace(/\s/g, "")

const moveDirs: Record<string, Direction> = {
  "^": [-1, 0],
  ">": [0, 1],
  "<": [0, -1],
  "v": [1, 0]
}

let startPos: Position = [0, 0]
gridArr.forEach((line, y) => {
  line.forEach((item, x) => {
    if (item === "@") startPos = [y, x]
  })
})

const moveRobot = (grid: Grid, pos: Position, move: string): [Grid, Position] => {
  const [dy, dx] = moveDirs[move]
  const [rY, rX] = pos

  const dirItems: string[] = []
  let i = 0
  while (true) {
    dirItems.push(grid[rY + i * dy][rX + i * dx])
    if (dirItems[dirItems.length - 1] === "#" || grid[rY + i * dy][rX + i * dx] === ".") {
      break
    }
    i++
  }

  if (dirItems[dirItems.length - 1] === "#") {
    return [grid, pos]
  }

  for (let j = 0; j < dirItems.length; j++) {
    grid[rY + j * dy][rX + j * dx] = dirItems[(j - 1 + dirItems.length) % dirItems.length]
  }

  return [grid, [rY + dy, rX + dx]]
}

let pos = startPos
for (const move of movesStr) {
  [gridArr, pos] = moveRobot(gridArr, pos, move)
}

console.log(
  gridArr.reduce((sum, line, y) => 
    sum + line.reduce((rowSum, item, x) => 
      rowSum + (item === "O" ? 100 * y + x : 0), 0), 0)
)

const wideMap: Record<string, [string, string]> = {
  "#": ["#", "#"],
  ".": [".", "."],
  "O": ["[", "]"],
  "@": ["@", "."]
}

const newGrid: string[][] = data[0].split("\n").map(line => 
  [...line].flatMap(char => wideMap[char]))

startPos = [0, 0]
newGrid.forEach((line, y) => {
  line.forEach((item, x) => {
    if (item === "@") startPos = [y, x]
  })
})

pos = startPos

const horiz: [number, number][] = [[0, 1], [0, -1]]

const search = (grid: string[][], pos: Position, move: string): [boolean, Set<Position>] => {
  const [dy, dx] = moveDirs[move]
  const [y, x] = pos
  const visited = new Set<string>()

  const dfs = (y: number, x: number): [boolean, Set<Position>] => {
    if (!(0 <= y && y < grid.length && 0 <= x && x < grid[0].length)) {
      return [true, new Set()]
    }
    if (grid[y][x] === "#") {
      return [true, new Set()]
    }
    if (grid[y][x] === ".") {
      return [false, new Set([[y, x]])]
    }
    if (visited.has(`${y},${x}`)) {
      return [false, new Set()]
    }

    visited.add(`${y},${x}`)
    const toMove = new Set<Position>()

    if (horiz.some(([hdy, hdx]) => hdy === dy && hdx === dx)) {
      const [stopped, positions] = dfs(y + dy, x + dx)
      if (stopped) {
        return [true, new Set()]
      }
      toMove.add([y, x])
      positions.forEach(pos => toMove.add(pos))
    } else {
      if (grid[y][x] === "[") {
        const [stopped, rightPos] = dfs(y, x + 1)
        if (stopped) {
          return [true, new Set()]
        }
        rightPos.forEach(pos => toMove.add(pos))
      } else if (grid[y][x] === "]") {
        const [stopped, leftPos] = dfs(y, x - 1)
        if (stopped) {
          return [true, new Set()]
        }
        leftPos.forEach(pos => toMove.add(pos))
      }

      const [stopped, nextPos] = dfs(y + dy, x)
      if (stopped) {
        return [true, new Set()]
      }

      toMove.add([y, x])
      nextPos.forEach(pos => toMove.add(pos))
    }

    return [false, toMove]
  }

  const [nextY, nextX] = [y + dy, x + dx]
  const [stopped, positions] = dfs(nextY, nextX)
  if (stopped) {
    return [false, new Set()]
  }
  return [true, positions]
}

const moveTwo = (grid: string[][], pos: Position, move: string): [string[][], Position] => {
  const [dy, dx] = moveDirs[move]
  const [ry, rx] = pos
  const [canMove, toMove] = search(grid, pos, move)
  if (!canMove) {
    return [grid, pos]
  }

  const positions = Array.from(toMove)
  if (dx > 0) {
    positions.sort(([y1, x1], [y2, x2]) => x2 - x1)
  } else if (dx < 0) {
    positions.sort(([y1, x1], [y2, x2]) => x1 - x2)
  } else if (dy > 0) {
    positions.sort(([y1, x1], [y2, x2]) => y2 - y1)
  } else {
    positions.sort(([y1, x1], [y2, x2]) => y1 - y2)
  }

  positions.forEach(([y, x]) => {
    if (toMove.has([y - dy, x - dx] as Position)) {
      grid[y][x] = grid[y - dy][x - dx]
    } else {
      grid[y][x] = "."
    }
  })

  grid[ry][rx] = "."
  grid[ry + dy][rx + dx] = "@"

  return [grid, [ry + dy, rx + dx]]
}

for (const move of movesStr) {
  [newGrid, pos] = moveTwo(newGrid, pos, move)
}

console.log(
  newGrid.reduce((sum, line, y) => 
    sum + line.reduce((rowSum, item, x) => 
      rowSum + (item === "[" ? 100 * y + x : 0), 0), 0)
)